interp/memory.go

```diff
@@ -0,0 +1,1392 @@
+	// function. They are stored here in a list so they can be removed if the
+	// execution of the function needs to be rolled back.
+	instructions []llvm.Value
+}
+
+// extend integrates the changes done by the sub-memoryView into this memory
+// view. This happens when a function is successfully interpreted and returns to
+// the parent, in which case all changed objects should be included in this
+// memory view.
+func (mv *memoryView) extend(sub memoryView) {
+	if mv.objects == nil && len(sub.objects) != 0 {
+		mv.objects = make(map[uint32]object)
+	}
+	for key, value := range sub.objects {
+		mv.objects[key] = value
+	}
+	mv.instructions = append(mv.instructions, sub.instructions...)
+}
+
+// revert undoes changes done in this memory view: it removes all instructions
+// created in this memoryView. Do not reuse this memoryView.
+func (mv *memoryView) revert() {
+	// Erase instructions in reverse order.
+	for i := len(mv.instructions) - 1; i >= 0; i-- {
+		llvmInst := mv.instructions[i]
+		if llvmInst.IsAInstruction().IsNil() {
+			// The IR builder will try to create constant versions of
+			// instructions whenever possible. If it does this, it's not an
+			// instruction and thus shouldn't be removed.
+			continue
+		}
+		llvmInst.EraseFromParentAsInstruction()
+	}
+}
+
+// markExternalLoad marks the given LLVM value as having an external read. That
+// means that the interpreter can still read from it, but cannot write to it as
+// that would mean the external read (done at runtime) reads from a state that
+// would not exist had the whole initialization been done at runtime.
+func (mv *memoryView) markExternalLoad(llvmValue llvm.Value) {
+	mv.markExternal(llvmValue, 1)
+}
+
+// markExternalStore marks the given LLVM value as having an external write.
+// This means that the interpreter can no longer read from it or write to it, as
+// that would happen in a different order than if all initialization were
+// happening at runtime.
+func (mv *memoryView) markExternalStore(llvmValue llvm.Value) {
+	mv.markExternal(llvmValue, 2)
+}
+
+// markExternal is a helper for markExternalLoad and markExternalStore, and
+// should not be called directly.
+func (mv *memoryView) markExternal(llvmValue llvm.Value, mark uint8) {
+	if llvmValue.IsUndef() || llvmValue.IsNull() {
+		// Undef pointer, nothing to mark.
+		return
+	}
+	if !llvmValue.IsAInstruction().IsNil() || !llvmValue.IsAArgument().IsNil() {
+		// These are considered external by default, there is nothing to mark.
+		return
+	}
+	global := getUnderlyingGlobal(llvmValue)
+	if !global.IsAGlobalVariable().IsNil() {
+		objectIndex := mv.r.getValue(global).(pointerValue).index()
+		obj := mv.get(objectIndex)
+		if obj.marked < mark {
+			obj = obj.clone()
+			obj.marked = mark
+			if mv.objects == nil {
+				mv.objects = make(map[uint32]object)
+			}
+			mv.objects[objectIndex] = obj
+		}
+		// TODO: also mark objects referenced by this object.
```

Comment:
Is this going to break anything now?

Reply:
In my testing, no. However it is technically unsound and should probably be fixed.

Reply:
So, does this require additional work now, or should we merge and deal with in future effort?

Reply:
Yes, I think I shouldn't be lazy and fix this (even if it doesn't seem to affect the behavior right now).

Reply:
Fixed now.